nx/lib/nx.ex

```diff
@@ -8451,6 +8451,43 @@ defmodule Nx do
     )
   end
 
+  ## Sigils
+
+  if Version.match?(System.version(), ">= 1.13.0-dev") do
+    defmacro sigil_M({:<<>>, _meta, [string]}, modifiers) do
```

Comment:
Nice work! 

This implementation only supports a literal tensor, without resolving interpolations. Is this intentional?
Perhaps a better error could be raised in such case. For now the behavior is:

```elixir
iex(7)> x = 1       
1
iex(8)> ~M[#{x} 2]
** (ArgumentError) cannot infer the numerical type of nil
    (nx 0.1.0-dev) lib/nx/type.ex:110: Nx.Type.infer/2
    (elixir 1.13.0-dev) lib/enum.ex:2394: Enum."-reduce/3-lists^foldl/2-0-"/3
    (nx 0.1.0-dev) lib/nx/type.ex:98: Nx.Type.infer/1
    (nx 0.1.0-dev) expanding macro: Nx.sigil_M/2
    iex:8: (file)
```

Reply:
Yes, this is intentional. I followed the convention in [built-in sigils](https://hexdocs.pm/elixir/1.12/Kernel.html#sigil_C/2) and used uppercase "M" to indicate this behaviour.

Reply:
I would try to raise a more semantic message then. Otherwise, this seems good :)

Reply:
It seems that for non-numerical values the current message is fine:

```elixir
iex(11)> ~M[1 2 "a" 3]
** (ArgumentError) cannot infer the numerical type of "a"
    (nx 0.1.0-dev) lib/nx/type.ex:110: Nx.Type.infer/2
    (elixir 1.13.0-dev) lib/enum.ex:2394: Enum."-reduce/3-lists^foldl/2-0-"/3
    (nx 0.1.0-dev) lib/nx/type.ex:98: Nx.Type.infer/1
    (nx 0.1.0-dev) expanding macro: Nx.sigil_M/2
    iex:11: (file)
```

Reply:
I'm not sure if this is also a predicted edge case:

```elixir
iex(11)> x = 1
1
iex(12)> ~M[1 2 x 3]  
warning: variable "x" does not exist and is being expanded to "x()", please use parentheses to remove the ambiguity or change the variable name
  nofile:1

** (CompileError) nofile:1: undefined function x/0 (there is no such import)
    (elixir 1.13.0-dev) lib/code.ex:404: Code.validated_eval_string/3
    (nx 0.1.0-dev) lib/nx.ex:8486: anonymous fn/1 in Nx.binary_to_numbers/1
    (elixir 1.13.0-dev) lib/enum.ex:1591: Enum."-map/2-lists^map/1-0-"/2
    (elixir 1.13.0-dev) lib/enum.ex:1591: Enum."-map/2-lists^map/1-0-"/2
    (nx 0.1.0-dev) lib/nx.ex:8486: anonymous fn/2 in Nx.binary_to_numbers/1
    (nx 0.1.0-dev) expanding macro: Nx.sigil_M/2
```

It seems that the sigil is trying to expand the variable on a "raw" env instead of the caller env

Reply:
Perhaps instead of using Code.eval_string, you could try to parse it with Integer.parse and Float.parse like so:

```elixir
defp binary_to_numbers(string) do
      for row <- String.split(string, "\n", trim: true) do
        row
        |> String.split(" ", trim: true)
        |> Enum.map(fn str ->
          case Integer.parse(str) do
            {int, ""} ->
              int

            _ ->
              case Float.parse(str) do
                {float, ""} -> float
                _ -> raise ArgumentError, "expected a numerical value for tensor, got #{str}"
              end
          end
        end)
      end
    end
```

This conditional structure is perhaps not ideal, but you can take this as a starting point. Maybe a `with` statement could be of use here.

Reply:
I tried to avoid it:
```elixir
defp binary_to_numbers(string) do
  for row <- String.split(string, "\n", trim: true) do
    row
    |> String.split(" ", trim: true)
    |> Stream.map(&Code.eval_string/1)
    |> Enum.map(fn
      {number, _} when is_number(number) -> number
      _ -> raise ArgumentError, "non-numerics and interpolations are not allowed"
    end)
  end
end
```

But there was another bad side effect:
```
iex(3)> ~M[x 2]
warning: variable "x" does not exist and is being expanded to "x()", please use parentheses to remove the ambiguity or change the variable name
  nofile:1
```

So I'll follow your advice, thanks :)

Reply:
I think I found a more concise way